certbot-apache/certbot_apache/augeasparser.py

```diff
@@ -0,0 +1,192 @@
+""" Tests for ParserNode interface """
+from certbot_apache import interfaces
+
+from acme.magic_typing import Dict, Tuple  # pylint: disable=unused-import, no-name-in-module
+
+
+class AugeasCommentNode(interfaces.CommentNode):
+    """ Augeas implementation of CommentNode interface """
+    ancestor = None
+    comment = ""
+    dirty = False
+    _metadata = dict()  # type: Dict[str, object]
+
+    def __init__(self, comment, ancestor=None):
+        self.comment = comment
+        self.ancestor = ancestor
+
+    def save(self, msg):  # pragma: no cover
+        pass
+
+    # Apache specific functionality
+
+    def get_metadata(self, key):
+        """ Returns a metadata object
+
+        :param str key: Metadata object name to return
+        :returns: Requested metadata object
+        """
+        try:
+            return self._metadata[key]
+        except KeyError:
+            return None
+
+class AugeasDirectiveNode(interfaces.DirectiveNode):
+    """ Augeas implementation of DirectiveNode interface """
+    ancestor = None
+    parameters = tuple()  # type: Tuple[str, ...]
+    dirty = False
+    enabled = True
+    name = ""
+    _metadata = dict()  # type: Dict[str, object]
+
+    def __init__(self, name, parameters=tuple(), ancestor=None):
+        self.name = name
+        self.parameters = parameters
+        self.ancestor = ancestor
+
+    def save(self, msg):  # pragma: no cover
+        pass
+
+    def set_parameters(self, parameters):  # pragma: no cover
+        self.parameters = tuple("CERTBOT_PASS_ASSERT")
+
+    # Apache specific functionality
+
+    def get_filename(self):
+        """Returns the filename where this directive exists on disk
+
+        :returns: File path to this node.
+        :rtype: str
+        """
+
+        # Following is the real implementation when everything else is in place:
+        # return apache_util.get_file_path(
+        #    self.parser.aug.get("/augeas/files%s/path" % apache_util.get_file_path(path)))
+        return "CERTBOT_PASS_ASSERT"
+
+    def get_metadata(self, key):
+        """ Returns a metadata object
+
+        :param str key: Metadata object name to return
+        :returns: Requested metadata object
+        """
+        try:
+            return self._metadata[key]
+        except KeyError:
+            return None
+
+    def has_parameter(self, parameter, position=None):
```

Comment:
I like it! I'm going to migrate this to mixin class.

Reply:
Another idea here is to forgo this in favor of something like `lower_parameters()` that simply returns all parameters in lowercase. This allows us to write:
```
parameter.lower() == node.lower_parameters()[position]
```
or
```
parameter.lower() in node.lower_parameters()
```
This could still be a method defined in a mixin class.

I think you should do what you want with this comment (including ignore it if you so choose), but part of me wants to figure out a way to be able to use simple Python syntax like `in` someway. It seems nice.